hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/types/HazelcastTypeCoercion.java

```diff
@@ -0,0 +1,422 @@
+        SqlNodeList thenOperands = call.getThenOperands();
+
+        List<SqlNode> operands = new ArrayList<>(thenOperands.size() + 1);
+        operands.addAll(thenOperands.getList());
+        operands.add(call.getElseOperand());
+
+        RelDataType[] types = inferTypes(binding.getScope(), operands, false);
+        if (types == null) {
+            return false;
+        }
+
+        // Do the coercion.
+
+        boolean coerced = false;
+        for (int i = 0; i < operands.size() - 1; ++i) {
+            coerced |= coerceElementType(binding.getScope(), thenOperands, i, types[i]);
+        }
+        coerced |= coerceOperandType(binding.getScope(), call, 3, types[operands.size() - 1]);
+
+        updateInferredType(call, types[types.length - 1]);
+
+        return coerced;
+    }
+
+    @Override
+    public RelDataType implicitCast(RelDataType in, SqlTypeFamily expected) {
+        // enables implicit conversion from CHAR to BOOLEAN
+        if (CHAR_TYPES.contains(typeName(in)) && expected == SqlTypeFamily.BOOLEAN) {
+            return TYPE_FACTORY.createSqlType(BOOLEAN, in.isNullable());
+        }
+
+        return super.implicitCast(in, expected);
+    }
+
+    @Override
+    protected void updateInferredType(SqlNode node, RelDataType type) {
+        ((HazelcastSqlValidator) validator).setKnownNodeType(node, type);
+        super.updateInferredType(node, type);
+    }
+
+    @Override
+    protected boolean coerceOperandType(SqlValidatorScope scope, SqlCall call, int index, RelDataType to) {
+        SqlNode operand = call.getOperandList().get(index);
+        RelDataType from = validator.deriveType(scope, operand);
+
+        // Just update the inferred type if casting is not needed. But if casting
+        // is not possible, still insert the cast to fail on its validation later.
+        if (!needToCast(scope, operand, to) && canCast(from, to)) {
+            updateInferredType(operand, to);
+            return false;
+        }
+
+        SqlNode cast = makeCast(operand, to);
+        call.setOperand(index, cast);
+        // derive the type of the newly created CAST immediately
+        validator.deriveType(scope, cast);
+        return true;
+    }
+
+    @SuppressWarnings("checkstyle:NPathComplexity")
+    @Override
+    protected boolean needToCast(SqlValidatorScope scope, SqlNode node, RelDataType to) {
+        RelDataType from = validator.deriveType(scope, node);
+
+        if (typeName(from) == typeName(to)) {
+            // already of the same type
+            return false;
+        }
+
+        if (typeName(from) == NULL || SqlUtil.isNullLiteral(node, false)) {
+            // never cast NULLs, just assign types to them
+            return false;
+        }
+
+        if (typeName(to) == ANY) {
+            // all types can be implicitly interpreted as ANY
+            return false;
+        }
+        if (typeName(from) == ANY) {
+            // casting from ANY is always required
+            return true;
```

Comment:
The branch is never invoked in tests.

Reply:
fixed in 30cedded5e3cd2372e5487b002b24eea87e9d35a